### DISCUSSION TITLE
Vm.run_func() FuncTypeMismatch using wasmedge_bindgen

### DISCUSSION THREAD

**Discussion by KeeganJin** [WasmEdge/WasmEdge#2829] [Q&A] [answered]
Hi all, 
I transformed the .wasm file of the [wasi-nn example](https://github.com/second-state/wasmedge-rustsdk-examples/tree/main/object-detection-via-wasinn) into a lib file and compiled it into .wasm file. It basically just changed the function name **main()** of [the pytorch example](https://github.com/second-state/WasmEdge-WASINN-examples/blob/master/pytorch-mobilenet-image/rust/src/main.rs) into **pytorch_infer()**, turn it into lib.rs and compiled it into wasm, so we can potentially use **wasmedge_bindgen** to get the inference result from the host. 

```
//lib.rs
#[wasmedge_bindgen]
pub fn pytorch_infer() {
    let args: Vec<String> = env::args().collect();
    let model_bin_name: &str = &args[1];
    let image_name: &str = &args[2];
     ----
   // no return yet
    )
```
But when I call the wasm-lib from the host by using
**vm.run_func(Some("extern"), "pytorch_infer", params!())?;** to call the wasm lib
(the rest part of the code is the same as the main.rs in [wasi-nn example](https://github.com/second-state/wasmedge-rustsdk-examples/tree/main/object-detection-via-wasinn) )
```
//main.rs
fn main(){
    infer()
}
fn infer(){
    let dir_mapping = &String::from(".:.");
    let wasm_file = &String::from("ml_pytorch_lib.wasm");
    ----


    vm.run_func(Some("extern"), "pytorch_infer", params!())?;
}
```
I got the error
```
load plugin
[2023-09-19 20:20:28.849] [error] execution failed: function signature mismatch, Code: 0x83
[2023-09-19 20:20:28.849] [error]     Mismatched function type. Expected: FuncType {params{i32 , i32} returns{i32}} , Got: FuncType {params{} returns{i32}}
Error: Core(Execution(FuncTypeMismatch))
```

So I am wondering how to make wasmedge_bindgen work in this case and how can the host function get the inference result from the wasmedge VM?

**Comment by apepkuss** [comment]
@KeeganJin If you'd like to use `wasmedge_bindgen` in your project, you should use [VmDocker](https://wasmedge.github.io/wasmedge-rust-sdk/wasmedge_sdk/dock/struct.VmDock.html) instead of Vm. This [example](https://github.com/second-state/wasmedge-rustsdk-examples/tree/main/wasmedge-bindgen) shows how to use VmDocker. It can help you. Thanks!

**Reply by KeeganJin** [reply]
Hi @apepkuss , thanks for your reply. I checked the documents and transformed Vm into VmDocker by doing 

```
    vm.wasi_module_mut()
        // .ok_or("Not found wasi module")?
        .expect("Not found wasi module")
        .initialize(
            Some(vec![wasm_file, model_bin, image_file]),
            None,
            Some(vec![dir_mapping]),
        );

    let vm = VmDock::new(vm);

// call the pytorch_infer in ml_pytorch_lib
    vm.run_func("pytorch_infer", params!())?;
```
And got an error saying that 
```
Error: Vm(NotFoundActiveModule)

```
Thanks. :>

**Comment by apepkuss** [comment]
From the error message, you didn't register the wasm module, which holds the `pytorch_infer` function, into the vm. You can use [Module::from_file](https://wasmedge.github.io/wasmedge-rust-sdk/wasmedge_sdk/struct.Module.html#method.from_file) and [Vm::register_module](https://wasmedge.github.io/wasmedge-rust-sdk/wasmedge_sdk/struct.Vm.html#method.register_module) to register the wasm file into the vm. After that, you're able to call `run_func` method. For example,

```rust
use wasmedge_sdk::Module;

// load wasm module that holds the `pytorch_infer` func
let module = Module::from_file(None, "/path/to/your/some.wasm")?;

// register the module as an active module into the vm
let vm = vm.register_module(None, module)?;
     
vm.wasi_module_mut()
        .expect("Not found wasi module")
        .initialize(
            Some(vec![wasm_file, model_bin, image_file]),
            None,
            Some(vec![dir_mapping]),
        );

let vm = VmDock::new(vm);
vm.run_func("pytorch_infer", params!())?;

```

In addition, I drafted a new [example](https://github.com/second-state/wasmedge-rustsdk-examples/tree/ggml-llama-example/ggml-llama-via-wasinn) for the `wasi_nn-ggml` plugin. It contains two parts: one is a wasm lib containing the `infer` function, the other an app that loads the wasm lib and invoke the `infer` func. Even though it does not use `wasmedge_bindgen`, the example should help you design your code.

**Reply by KeeganJin** [reply]
Hi @apepkuss ,

thanks for your reply.
In the previous code, I registered the WASM file by doing
```
   // build the Vm
    let mut vm = VmBuilder::new()
        .with_config(config)
        .with_plugin_wasi_nn()
        .build()?
        .register_module(Some("extern"), module)?;
```

The complete code is as shown below, thanks:

```
#[cfg(all(target_os = "linux", target_arch = "x86_64"))]
use wasmedge_sdk::{
    config::{CommonConfigOptions, ConfigBuilder, HostRegistrationConfigOptions},
    params,
    dock::{VmDock},
    plugin::PluginManager,
    Module, VmBuilder,
};

struct InferenceResult(usize, f32);

fn main() -> Result<(), Box<dyn std::error::Error>> {
    #[cfg(all(target_os = "linux", target_arch = "x86_64"))]
    infer()?;
    println!("first inference");

    // println!("second inference");
    // infer()?;

    Ok(())
}


#[cfg(all(target_os = "linux", target_arch = "x86_64"))]
fn infer() -> Result<(), Box<dyn std::error::Error>> {

    //works in the Terminal
    let dir_mapping = &String::from(".:.");
    // let wasm_file = &String::from("wasmedge-wasinn-example-mobilenet-image.wasm");
    // let wasm_file = &String::from("ml_pytorch_lib.wasm");
    let wasm_file = &String::from("ml_pytorch.wasm");

    let model_bin = &String::from("mobilenet.pt");
    let image_file = &String::from("input.jpg");
    // let args: Vec<String> = vec![dir_mapping,wasm_file,model_bin,image_file];


    println!("load plugin");
    PluginManager::load(None)?;

    let config = ConfigBuilder::new(CommonConfigOptions::default())
        .with_host_registration_config(HostRegistrationConfigOptions::default().wasi(true))
        .build()?;
    assert!(config.wasi_enabled());
    // assert!(config.wasi_nn_enabled());

    // load wasm module from file
    let module = Module::from_file(Some(&config), wasm_file)?;

   // build a Vm
    let mut vm = VmBuilder::new()
        .with_config(config)
        .with_plugin_wasi_nn()
        .build()?
        .register_module(Some("extern"), module)?;


    // init wasi module
    vm.wasi_module_mut()
        // .ok_or("Not found wasi module")?
        .expect("Not found wasi module")
        .initialize(
            Some(vec![wasm_file, model_bin, image_file]),
            None,
            Some(vec![dir_mapping]),
        );
    
    // vm.run_func(Some("extern"), "_start", params!())?;

    let vm = VmDock::new(vm);

    // call the call_infer in ml_pytorch_lib.wasm
    vm.run_func("call_infer", params!())?;


    // vm.run_func(Some("extern"), "_start", params!())?;

    Ok(())
}

```

**Comment by apepkuss** [comment, marked as answer]
As your code shows, you register `module` as a named module. It is named "extern". However, you invoke the `call_infer` from an active module that does not exist. 

Fix: replace `.register_module(Some("extern"), module)?;` with `.register_module(None, module)?;`, then try again.

```rust
...

#[cfg(all(target_os = "linux", target_arch = "x86_64"))]
fn infer() -> Result<(), Box<dyn std::error::Error>> {

    ...

    // load wasm module from file
    let module = Module::from_file(Some(&config), wasm_file)?;

   // build a Vm
    let mut vm = VmBuilder::new()
        .with_config(config)
        .with_plugin_wasi_nn()
        .build()?
        .register_module(Some("extern"), module)?;   <=========== register a named module, called "extern"



   ...

    let vm = VmDock::new(vm);

    // call the call_infer in ml_pytorch_lib.wasm
    vm.run_func("call_infer", params!())?;           <============ call "call_infer" from an active module that doesn't exist


    // vm.run_func(Some("extern"), "_start", params!())?;

    Ok(())
}

```

**Reply by KeeganJin** [reply]
Thanks a lot! That fixes that issue!

**Reply by apepkuss** [reply]
If the issue have been solved, please close it. Thanks a lot!
